Documentation/project-docs/AssemblyBlobs.md

```diff
@@ -0,0 +1,214 @@
+        uint32_t data_size;
+        uint32_t debug_data_offset;
+        uint32_t debug_data_size;
+        uint32_t config_data_offset;
+        uint32_t config_data_size;
+    };
+
+Only the `data_offset` and `data_size` fields must have a non-zero
+value, other fields describe optional data and can be set to `0`. 
+
+Individual fields have the following meanings:
+
+  - `data_offset`: offset of the assembly image data from the
+    beginning of the blob file
+  - `data_size`: number of bytes of the image data
+  - `debug_data_offset`: offset of the assembly's debug data from the
+    beginning of the blob file. A value of `0` indicates there's no
+    debug data for this assembly.
+  - `debug_data_size`: number of bytes of debug data. Can be `0` only
+    if `debug_data_offset` is `0`
+  - `config_data_offset`: offset of the assembly's config file data
+    from the  beginning of the blob file. A value of `0` indicates
+    there's no config file data for this assembly.
+  - `config_data_size`: number of bytes of config file data. Can be
+    `0` only if `config_data_offset` is `0`
+
+## Index blob
+
+Each application will contain exactly one blob with a global index -
+two tables with assembly name hashes.  All the other blobs **do not**
+contain these tables.  Two hash tables are necessary because hashes
+for 32-bit and 64-bit devices are different.
+
+The hash tables follow the [Assembly descriptor
+table](#assembly-descriptor-table) and precede the individual assembly
+streams.
+
+Placing the hash tables in a single index blob, while "wasting" a
+certain amount of memory (since 32-bit devices won't use the 64-bit
+table and vice versa), makes for simpler and faster runtime
+implementation and the amount of memory wasted isn't big (1000
+two tables which are 8kb long each, this being the amount of memory
+wasted)
+
+### Hash table format
+
+Both tables share the same format, despite the hashes themselves being
+of different sizes.  This is done to make handling of the tables
+easier on the runtime.
+
+Each entry contains, among other fields, the assembly name hash.  The
+hash value is obtained using the
+[xxHash](https://cyan4973.github.io/xxHash/) algorithm and is
+calculated **without** including the `.dll` extension.  This is done
+for runtime efficiency as the vast majority of Mono requests to load
+an assembly does not include the `.dll` suffix, thus saving us time of
+appending it in order to generate the hash for index lookup.
+
+Each entry is represented by the following structure:
+
+    struct BlobHashEntry
+    {
+        union {
+            uint64_t hash64;
+            uint32_t hash32;
+        };
+        uint32_t mapping_index;
+        uint32_t local_blob_index;
+        uint32_t blob_id;
+    };
+
+Individual fields have the following meanings:
+
+ - `hash64`/`hash32`: the 32-bit or 64-bit hash of the assembly's name
+   **without** the `.dll` suffix
+ - `mapping_index`: index into a compile-time generated array of
```

Comment:
I'm not sure what this means, but strongly implies that the App needs to "know about" Features.  This worries me?